torch/distributed/nn/jit/templates/remote_module_template.py

```diff
@@ -13,13 +13,47 @@
 
 
 {jit_script_decorator}
-def _remote_forward(module_rref: RRef[module_interface_cls], {arg_types}){arrow_and_return_type}:
+def _remote_forward(
+    module_rref: RRef[module_interface_cls], device: str, {arg_types}){arrow_and_return_type}:
     module = module_rref.local_value()
-    return module.forward({args}, {kwargs})
+    device = torch.device(device)
+
+    if device.type != "cuda":
+        return module.forward({args}, {kwargs})
+
+    # If the module is on a cuda device,
+    # move any CPU tensor in args or kwargs to the same cuda device.
+    # Since torch script does not support generator expression,
+    # have to use tuple concatenation instead of
+    # ``tuple(i.to(device) if isinstance(i, Tensor) else i for i in *args)``.
+    args = ({args},)
+    out_args = ()
+    for arg in args:
+        arg = (arg.to(device),) if isinstance(arg, Tensor) else (arg,)
```

Comment:
What if `args` contains data structures with nested tensors, such as `List[tensor]` or `Dict[str, tensor]`? IIUC torchscript supports those types so we may want to extract tensors from them and move those too?  Although this is an enhancement and can probably be saved for a follow up PR

Reply:
Good point! A warning is added here.

I am not sure if there is any real use case that passes nested tensors in args or kwargs. Have you seen such an example in a real use case? If not, I prefer not overcomplicate the implementation at this time.

Reply:
I have encountered a few use cases (e.g. internal hpc workload) where the return type contains `List[Tensor]` so users may hit this when training more complex model. But agreed, I don't think it's a hard blocker on this PR.

Worth mentioning in passing - it is actually quite easy to unpack/repack tensors from nested data structure, see pytree utils: https://github.com/pytorch/pytorch/blob/master/torch/utils/_pytree.py